test/AppNotificationTests/AppNotificationHelpers.h

```diff
@@ -0,0 +1,237 @@
+            break;
+
+        default:
+            // Unexpected selection
+            VERIFY_FAIL();
+        }
+    }
+
+    bool VerifyToastsPosted(std::vector<winrt::Microsoft::Windows::AppNotifications::AppNotification> const& expectedToastVector)
+    {
+        bool result{ false };
+        for (int i { 0 }; i < 5; i++)
+        {
+            auto getAllAsync{ winrt::Microsoft::Windows::AppNotifications::AppNotificationManager::Default().GetAllAsync() };
+            auto scopeExitGetAll = wil::scope_exit(
+                [&] {
+                    getAllAsync.Cancel();
+                });
+
+            VERIFY_ARE_EQUAL(getAllAsync.wait_for(std::chrono::milliseconds(500)), winrt::Windows::Foundation::AsyncStatus::Completed);
+            scopeExitGetAll.release();
+
+            auto actualToastVector{ getAllAsync.GetResults() };
+            if (actualToastVector.Size() == expectedToastVector.size())
+            {
+                for (int actualToastIndex{ 0 }; (uint32_t)actualToastIndex < actualToastVector.Size(); actualToastIndex++)
+                {
+                    for (int expectedToastIndex{ 0 }; expectedToastIndex < expectedToastVector.size(); expectedToastIndex++)
+                    {
+                        if (actualToastVector.GetAt(actualToastIndex).Id() == expectedToastVector[expectedToastIndex].Id())
+                        {
+                            VERIFY_ARE_EQUAL(actualToastVector.GetAt(actualToastIndex).Payload(), expectedToastVector[expectedToastIndex].Payload());
+                            break;
+                        }
+                    }
+                }
+
+                result = true;
+                break;
+            }
+
+            Sleep(500);
+        }
+        return result;
+    }
+
+    void PostToastHelper(std::wstring const& tag, std::wstring const& group)
+    {
+        winrt::AppNotification toast{ CreateToastNotification() };
+
+        toast.Tag(tag.c_str());
+        toast.Group(group.c_str());
+
+        winrt::AppNotificationManager::Default().Show(toast);
+
+        std::vector<winrt::AppNotification> toastVector{ toast };
+        VERIFY_IS_TRUE(VerifyToastsPosted(toastVector));
+    }
+
+    bool VerifyToastsCleared()
+    {
+        bool result{ false };
+        for (int i{ 0 }; i < 5; i++)
+        {
+            auto getAllAsync{ winrt::Microsoft::Windows::AppNotifications::AppNotificationManager::Default().GetAllAsync() };
+            auto scopeExitGetAll = wil::scope_exit(
+                [&] {
+                    getAllAsync.Cancel();
+                });
+
+            VERIFY_ARE_EQUAL(getAllAsync.wait_for(c_timeout), winrt::Windows::Foundation::AsyncStatus::Completed);
+            scopeExitGetAll.release();
+
+            auto actualToastVector{ getAllAsync.GetResults() };
+            if (actualToastVector.Size() == (uint32_t) 0)
+            {
+                result = true;
+                break;
+            }
+
+            Sleep(500);
```

Comment:
1. Magic constant
2. Why 500? Why not 100 or 5000?
3. Sleep is generally unwise as it imposes test overhead, and if used to try to ccoordinate cross-thread or -process behavior is an unreliable mechanism to do so.

Reply:
@hulumane What do you think we should do here? Currently the tests pass without the sleep but I'm unsure how to go about this.

Reply:
This code uses polling to determine that the ToastHistory ultimately is set to 0 within 2.5 seconds. We don't have a better way to do this because the Clear APIs are a fire and forget operation that do thing asynchronously. So if we call ClearAll, the call to VerifyToastsCleared() may not return true because of timing issues. Hence the polling.

As far as the 500ms number is concerned, we should make it 1000. That way this code polls 5 times in 5 seconds which is a little more understandable.

In the future, we could presumably make this an Async API and rely on a ClearAll event to signal completion of the Async operation. Alas, with our current OS setup, this isn't possible.

Please file a bug @pmpurifoy to have an Async equivalent for all our Toast APIs. We can poll our developers to see if this is something they want.

Reply:
https://github.com/microsoft/WindowsAppSDK/issues/2617

Reply:
> As far as the 500ms number is concerned, we should make it 1000.

Better to do actual timeout than sleep + N loops

```
const auto c_timeout{ 5s };
const auto c_delay{ 25ms };
auto start{ GetTickCount() };
auto elapsed{ start };
for (;;)
{
    IF do something and happy
        break

    elapsed = GetTickCount() - start;
    IF elapsed > c_timeout
        break
    Sleep(c_delay)
}
IF elapsed > c_timeout
    ERROR
```

Cheap timeout loop to ensure you run for X time with small sleeps while awaiting results

Sleeping for 1s is an awefully long time. Small sleeps give up time for plenty of action to happen on other threads (in this process and others) without burning the cpu in a spinloop. 10ms, 25ms and/or 100ms are usually good numbers for these short sleeps. YMMV